contributors/THIRD-PARTY-EXTENSION.md

```diff
@@ -0,0 +1,120 @@
+## Introduction
+
+Welcome to the guide for developing third party extensions in the `create-eth` repository.
+
+Third party extension allows developers to **extend** base instance created by `npx create-eth@latest` and publish the extension to github that can be used by other developers via:
+
+```bash
+npx create-eth@latest -e {your-github-userName}/{extension-repo-name}:{extension-branch-name} # extension-branch-name is optional
+```
+
+It is important to understand distinction between instance and extension.
+
+- Instance is created by `npx create-eth@latest` and contains full project files.
+- Extension does not contain full project files, instead it only contains extra files, extra directories and [`*.args.mjs`](TEMPLATING.md#args-files) files that will added to the base instance created via `npx create-eth@latest` with `-e` flag.
+
+> TIP: Checkout directory structure of `eip-712` extension maintained Scaffold-ETH team at https://github.com/scaffold-eth/create-eth-extensions/tree/eip-712
+
+This document outlines the workflow for creating and testing your extensions.
+
+## Workflow Overview
+
+An developer goes through 2 phases while developing extension:
+
+- **Phase 1**: Initial Setup and Development
+  - This phase helps you create your own extension based on the changes you make to instance created by `npx create-eth@latest`
+- **Phase 2**: Local Testing and Publishing
+  - This phase helps you to try out your extension locally and see how it works when used by other developers before publishing to github
```

Comment:
Question about these two phases. Let's say I start my initial setup and development with Phase 1, I added a new page and I use the `yarn create-extension` command to copy files into the `externalExtensions` folder.

If I then decide I want to edit my new page, or rename my page, should I edit it and run the `yarn create-extension` command again, or should I move onto Phase Two to do these edits?

Is there a preference between running the `yarn create-extension` command multiple times versus moving onto Phase Two and using the `--dev` method to make edits?

Reply:
Ohh yeah same doubt popped into my head a while ago and maybe something we need to get more clear in docs. 

The difference b/w phase 1 and phase 2 is: 

If you move to phase 2(which basically create a symlinked instance based of your new extension + base SE-2) and lets suppose you want to add new page / file to your extension and in phase 2 you won't be able to do that.  Phase 2 only supports editing the content inside the generated instance files.

Another important use case of phase 2 is that it allow's other collaborators of your [extension](https://github.com/scaffold-eth/create-eth-extensions/tree/eip-712) to pull the extension code locally and test it out. Because before there was no way to test your created extension locally, instead you were needed to push the code blindly to your github repo and then test it via `npx create-eth@latest -e {yourGHUsername}/{yourExtension}`.


> Is there a preference between running the yarn create-extension command multiple times versus moving onto Phase Two and using the --dev method to make edits?


Not sure if this diagram helps : 


![image](https://github.com/user-attachments/assets/2a3200c3-fe4b-4585-ad12-8854ab927388)


Phase 1 and phase 2 requires creating two different instances, So its ergonomic to do as much as you can in phase 1. Because phase 1 contains all things which we could do in phase 2, just that in phase 1 you need to run `yarn create-extension` command again and again. 

> If I then decide I want to edit my new page, or rename my page, should I edit it and run the yarn create-extension command again, or should I move onto Phase Two to do these edits?

So changing the content inside file running `yarn create-extension` again and again is the way to go due to ergonomics and also after moving to phase 2 you won't be able to add any extra files. 


Regarding renaming the file name, this is bug/feature which we need to solve because currently in phase 1, let say you had a file name "Foo.tsx" in your instance. 
1. You ran `yarn create-extension` this copies `Foo.tsx` in your extension.
2. You renamed file `Foot.tsx` => `Bar.tsx` in instance and commit those changes.
3. Now if you run `yarn create-extension` this won't delete previous `Foot.tsx` file instead it will create new `Bar.tsx` file in the extension

Renaming file also doesn't seem to work in Phase 2(maybe that's how symlinks works they are gone when you rename a file?) 
 
Also cc @carletex and others to chime in and would love to know their opinions too / correct me if I missed something!

Reply:
@technophile-04 Got it, thank you for the thorough explanation!

Reply:
amazing diagram @technophile-04 <3

+1 on all you said.

> Another important use case of phase 2 is that it allow's other collaborators of your [extension](https://github.com/scaffold-eth/create-eth-extensions/tree/eip-712) to pull the extension code locally and test it out. Because before there was no way to test your created extension locally, instead you were needed to push the code blindly to your github repo and then test it via npx create-eth@latest -e {yourGHUsername}/{yourExtension}.

Exactly! This also makes me wonder if we should frame the dev workflow differently. Currently, we have phases 1 & 2... which signals that you have to do in this order, and you can't miss any steps.

But in reality, phase 1 is kind of optional (just a helper) and phase 2 is the way to test locally (you could even test directly in the repo, not very convenient but you could do it)

My concern is that it might seem more complicated than it really is. I realized this when explaining it to Hunter... when in reality it's just: "Create the files/folder with this structure (+ be aware of template files & deps)"

@ChangoMan let's see how unnatural it feels when you try to explain it on a video.

Let's keep iterating as we go on future PRs and see how it goes. 

Thanks all, good stuff